resctrl/collector.go

```diff
@@ -28,54 +29,67 @@ import (
 )
 
 type collector struct {
-	id               string
-	interval         time.Duration
-	getContainerPids func() ([]string, error)
-	resctrlPath      string
-	running          bool
+	id                string
+	interval          time.Duration
+	getContainerPids  func() ([]string, error)
+	resctrlPath       string
+	running           bool
+	numberOfNUMANodes int
+	mu                sync.Mutex
 }
 
-func newCollector(id string, getContainerPids func() ([]string, error), interval time.Duration) *collector {
-	return &collector{id: id, interval: interval, getContainerPids: getContainerPids}
+func newCollector(id string, getContainerPids func() ([]string, error), interval time.Duration, numberOfNUMANodes int) *collector {
+	return &collector{id: id, interval: interval, getContainerPids: getContainerPids, numberOfNUMANodes: numberOfNUMANodes,
+		mu: sync.Mutex{}}
 }
 
 func (c *collector) setup() error {
-	err := c.prepareMonGroup()
+	var err error
+	c.resctrlPath, err = prepareMonitoringGroup(c.id, c.getContainerPids)
 
-	if c.interval != 0 && c.id != rootContainer {
+	if c.interval != 0 {
 		if err != nil {
-			klog.Errorf("Failed to setup container %q resctrl collector: %v \n Trying again in next intervals!", c.id, err)
+			c.running = false
+			klog.Errorf("Failed to setup container %q resctrl collector: %w \n Trying again in next intervals.", c.id, err)
+		} else {
+			c.running = true
 		}
 		go func() {
 			for {
+				time.Sleep(c.interval)
+				c.mu.Lock()
+				klog.V(5).Infof("Trying to check %q containers control group.", c.id)
 				if c.running {
-					err = c.prepareMonGroup()
+					err = c.checkMonitoringGroup()
 					if err != nil {
-						klog.Errorf("checking %q resctrl collector but: %v", c.id, err)
+						c.running = false
+						klog.Errorf("Failed to check %q resctrl collector control group: %w \n Trying again in next intervals.", c.id, err)
 					}
 				} else {
-					err = c.clear()
+					c.resctrlPath, err = prepareMonitoringGroup(c.id, c.getContainerPids)
 					if err != nil {
-						klog.Errorf("trying to end %q resctrl collector interval but: %v", c.id, err)
+						c.running = false
+						klog.Errorf("Failed to setup container %q resctrl collector: %w \n Trying again in next intervals.", c.id, err)
+					} else {
+						c.running = true
 					}
-					break
 				}
-				time.Sleep(c.interval)
+				c.mu.Unlock()
 			}
 		}()
 	} else {
 		// There is no interval set, if setup fail, stop.
 		if err != nil {
-			c.running = false
-			return err
+			return fmt.Errorf("failed to setup container %q resctrl collector: %w", c.id, err)
 		}
+		c.running = true
```

Comment:
Looks like unserialised write to `c.running`.

Reply:
There is the mutex inside this func.